### PR TITLE
[VPlan] Initial modeling of VF * UF as VPValue.

### PR DESCRIPTION
**Pull request by fhahn**
This patch starts initial modeling of VF * UF in VPlan.
Initially, introduce a dedicated VFxUF VPValue, which is then
populated during VPlan::prepareToExecute. Initially, the VF * UF
applies only to the main vector loop region. Once we extend the
scope of VPlan in the future, we may want to associate different VFxUFs
with different vector loop regions (e.g. the epilogue vector loop)

This allows explicitly parameterizing recipes that rely on the
VF * UF, like the canonical induction increment. At the moment, this
mainly helps to avoid generating some duplicated calls to vscale with
scalable vectors. It should also allow using EVL as induction increments
explicitly in D99750. Referring to VF * UF is also needed in other
places that we plan to migrate to VPlan, like the minimum trip count
check during skeleton creation.

The first version creates the value for VF * UF directly in
prepareToExecute to limit the scope of the patch. A follow-on patch will
model VF * UF computation explicitly in VPlan using recipes.

Moved from Phabricator (https://reviews.llvm.org/D157322)

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 1]
```diff
@@ -1168,13 +1166,26 @@ class VPInstruction : public VPRecipeWithIRFlags, public VPValue {
...
     case VPInstruction::ActiveLaneMask:
     case VPInstruction::CalculateTripCountMinusVF:
-    case VPInstruction::CanonicalIVIncrement:
     case VPInstruction::CanonicalIVIncrementForPart:
     case VPInstruction::BranchOnCount:
       return true;
     };
     llvm_unreachable("switch should return");
   }
+
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlan.h:1175]: /// Returns true if the recipe only uses the first part of operand \p Op.

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 2]
```diff
@@ -2624,6 +2644,9 @@ class VPlan {
   /// The vector trip count.
   VPValue &getVectorTripCount() { return VectorTripCount; }
 
+  /// Returns runtime VF * UF for the vector loop region.
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlan.h:2647]:
```suggestion
  /// Returns the VF * UF of the vector loop region.
```

**Review by ayalz — APPROVED** [review]
LGTM, couple of minor nits.

File: llvm/lib/Transforms/Vectorize/VPlan.h (hunk 2 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlan.h]: Done, thanks!

File: llvm/lib/Transforms/Vectorize/VPlan.h (hunk 1 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlan.h:1175]: Added, thanks!

**Pull request merged**